lib/matplotlib/scale.py

```diff
@@ -94,6 +94,98 @@ def get_transform(self):
         return IdentityTransform()
 
 
+class FuncTransform(Transform):
+     """
+     A simple transform that takes and arbitrary function for the
+     forward and inverse transform.
+     """
+
+     input_dims = 1
+     output_dims = 1
+     is_separable = True
+     has_inverse = True
+
+     def __init__(self, forward, inverse):
+         """
+         Parameters
+         ----------
+
+         forward: callable
+             The forward function for the transform.  This function must have
+             an inverse and, for best behavior, be monotonic.
+             It must have the signature::
+
+                def forward(values: array-like) ->
+                        array-likeThe forward function for the transform
+
+         inverse: callable
+            The inverse of the forward function.  Signature as ``forward``.
+         """
+         super().__init__()
+         if callable(forward) and callable(inverse):
+             self._forward = forward
+             self._inverse = inverse
+         else:
+             raise ValueError('arguments to FuncTransform must '
+                              'be functions')
+
+     def transform_non_affine(self, values):
+         return self._forward(values)
+
+     def inverted(self):
+         return FuncTransform(self._inverse, self._forward)
+
+
+class FuncScale(ScaleBase):
+     """
+     Provide an arbitrary scale with user-supplied function for the axis.
+     """
+
+     name = 'function'
+
+     def __init__(self, axis, functions):
+         """
+         Parameters
+         ----------
+
+         axis: the axis for the scale
+
+         functions: (callable, callable)
+            two-tuple of the forward and inverse functions for the scale.
+            The forward function must have an inverse and, for best behavior,
+            be monotonic.
+
+            Both functions must have the signature::
+
+               def forward(values: array-like) ->
+                       array-likeThe forward function for the transform
```

Comment:
something wrong with the formatting

Reply:
oooops.   Should be

```
            Both functions must have the signature::

               def forward(values: array-like) returns array-like
```

Reply:
Actually
```
def forward(values: array-like) -> array-like
```
looks best to me (it's nearly normal type annotation syntax), but not insisting on that.